### ISSUE TITLE
Clear config list value from the command line

### ISSUE THREAD

**Issue by michellemay** [typesafehub/config#380]
In DefaultTransformer.java, it's not currently possible to clear the content of a config list from the command line parameter special syntax (ex: -Dkey.0=value).

For instance, if reference.conf have default values for a list, it's not possible to clear it's content.

**Comment by havocp** [comment]
Not just for lists - there's no way to clear any value. We get this information as a hash table (the system properties dict). We aren't parsing the command line or anything directly.

**Comment by michellemay** [comment]
I could imagine different solutions like: 
  -Dkey.empty
  -Dkey.0=
  -Dkey=

So when the DefaultTransformer see an object with such construct, it could still issue a new SimpleConfigList(value.origin(), emptyList);

**Comment by havocp** [comment]
empty string is also a valid value to want to set though :-/ I guess if you ask for a list and the value is an empty string we could convert that to an empty list, perhaps. (at the time of asking for the list)

**Comment by michellemay** [comment]
Exactly, that's what I was thinking.  From what I understand, this type mapping is already performed at query time.

My preferred solution would be '-Dkey='
However, for symmetry with the indexed syntax, I might find it better to have something in the key itself. (hence my other suggestions).  I agree that the '-Dkey.0=' is not a valid solution since we can have empty strings in lists too.

**Comment by michellemay** [comment]
The solution is to change the implementation of the transformer.  We could implement our own custom check for empty lists.  However, DefaultTransformer is hardcoded in SimpleConfig and SimpleConfig is final.  That would be a nice addition to the library to be able to subclass SimpleConfig and override some of the defaults like the transformer.

**Comment by havocp** [comment]
I think it's important to keep SImpleConfig private; it isn't properly thought out as a public API (starting with the class name which is bad), and it would greatly reduce our ability to evolve the library without breaking ABI if we made it public. But there may be other ways to achieve the same end.

I think we could do a PR that did empty string => empty list when you `Config.getList`, I can't think of anything that would break, though potentially the unit tests would shake out some problem. Non-empty string should still be an error in this case.

I think there was some past idea that you'd be able to set the transformer; that's why it's a separate class. But I may have found it didn't work out well in the details, or maybe I just never got around to it, I don't remember. It would probably need to be something like `ConfigParseOptions.setTransformer` and then SimpleConfig would need a transformer field in it. But I'm not sure it's really worth it.

**Comment by ILLKO** [comment]
No solution still, as the pull request got closed?

**Comment by havocp** [comment]
No, I don't think it's been solved.
